core/tests/protractor_utils/ExplorationEditorMainTab.js

```diff
@@ -0,0 +1,978 @@
+  };
+
+  // RULES
+  this.selectRuleInAddResponseModal = function(interactionId, ruleName) {
+    _selectRule(addResponseDetails, interactionId, ruleName);
+  };
+
+  var _getRuleDescription = function(interactionId, ruleName) {
+    if (ruleTemplates.hasOwnProperty(interactionId)) {
+      if (ruleTemplates[interactionId].hasOwnProperty(ruleName)) {
+        return ruleTemplates[interactionId][ruleName].description;
+      } else {
+        throw Error('Unknown rule: ' + ruleName);
+      }
+    } else {
+      throw Error('Could not find rules for interaction: ' + interactionId);
+    }
+  };
+
+  // Parses the relevant ruleDescription string, and returns an Array containing
+  // the types of the rule input parameters.
+  var _getRuleParameterTypes = function(interactionId, ruleName) {
+    var ruleDescription = _getRuleDescription(interactionId, ruleName);
+    // An example of rule description:
+    // is equal to {{a|NonnegativeInt}} and {{b|NonnegativeInt}}.
+    // (from NumericInput).
+    var parameterTypes = [];
+    var re = /\|(.*?)\}/ig;
+    // Matched result = Array[|NonnegativeInt}, |NonnegativeInt}]
+    var angularSelectors = ruleDescription.match(re);
+    // Slicing first and last letter.
+    if (angularSelectors) {
+      angularSelectors.forEach(function(elem) {
+        parameterTypes.push(elem.toString().slice(1, -1));
+      });
+    }
+    // Expected sample output = Array[NonnegativeInt, NonnegativeInt]
+    return parameterTypes;
+  };
+
+  // This function sets the parameter values for the given rule.
+  // Note: The parameter values should be specified as additional arguments
+  // after the ruleName. For example, the call
+  //   _selectRuleParameters(ruleElement, 'NumericInput', 'Equals', 24)
+  // will result in a rule that checks whether the learner's answer equals 24.
+  var _setRuleParameters = function(ruleElement, interactionId, ruleName) {
+    var parameterValues = [];
+    for (var i = 3; i < arguments.length; i++) {
+      parameterValues.push(arguments[i]);
+    }
+    var parameterTypes = _getRuleParameterTypes(interactionId, ruleName);
+    expect(parameterValues.length).toEqual(parameterTypes.length);
+    var answerDescriptionFragment = element.all(
+      by.css('.protractor-test-answer-description-fragment'));
+    for (var i = 0; i < parameterValues.length; i++) {
+      var parameterElement = answerDescriptionFragment.get(i * 2 + 1);
+      var parameterEditor = forms.getEditor(
+        parameterTypes[i])(parameterElement);
+
+      if (interactionId === 'MultipleChoiceInput') {
+      // This is a special case as it uses a dropdown to set a NonnegativeInt.
+        parameterElement.element(by.tagName('button')).click();
+        multipleChoiceAnswerOptions(parameterValues[i])
+          .click();
+      } else {
+        parameterEditor.setValue(parameterValues[i]);
+      }
+    }
+  };
+
+  /**
+   * Parse for Angular selectors in ruleDescription and replace them.
```

Comment:
I actually forgot to comment on this earlier -- these aren't Angular selectors, they are "rule input placeholders". They use a similar syntax but are conceptually different, and they don't get interpolated using Angular.

OK to fix in separate PR. Note that this affects naming etc. below as well.

Reply:
Renaming to ruleInputPlaceholders 👍